src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -40,6 +101,283 @@ public override bool CanConvert(Type typeToConvert)
+
+            bool success = OnTryWrite(writer, value, options, ref state);
+            if (success)
+            {
+                VerifyWrite(state.Current.OriginalDepth, writer);
+                // No need to clear state.Current.OriginalDepth since a stack pop will occur.
+            }
+
+            state.Pop(success);
+
+            return success;
+        }
+
+        internal bool TryWriteDataExtensionProperty(Utf8JsonWriter writer, T value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            Debug.Assert(this is JsonDictionaryConverter<T>);
+
+            if (writer.CurrentDepth >= options.EffectiveMaxDepth)
+            {
+                ThrowHelper.ThrowInvalidOperationException_SerializerCycleDetected(options.MaxDepth);
+            }
+
+            bool success;
+            JsonDictionaryConverter<T> dictionaryConverter = (JsonDictionaryConverter<T>)this;
+
+            if (ClassType == ClassType.Value)
+            {
+                Debug.Assert(!state.IsContinuation);
+
+                int originalPropertyDepth = writer.CurrentDepth;
+
+                // Ignore the naming policy for extension data.
+                state.Current.IgnoreDictionaryKeyPolicy = true;
+
+                success = dictionaryConverter.OnWriteResume(writer, value, options, ref state);
+                if (success)
+                {
+                    VerifyWrite(originalPropertyDepth, writer);
+                }
+            }
+            else
+            {
+                bool isContinuation = state.IsContinuation;
+
+                state.Push();
+
+                if (!isContinuation)
+                {
+                    Debug.Assert(state.Current.OriginalDepth == 0);
+                    state.Current.OriginalDepth = writer.CurrentDepth;
+                }
+
+                // Ignore the naming policy for extension data.
+                state.Current.IgnoreDictionaryKeyPolicy = true;
+
+                success = dictionaryConverter.OnWriteResume(writer, value, options, ref state);
+                if (success)
+                {
+                    VerifyWrite(state.Current.OriginalDepth, writer);
+                }
+
+                state.Pop(success);
+            }
+
+            return success;
+        }
+
+        internal override sealed Type TypeToConvert => typeof(T);
+
+        internal void VerifyRead(JsonTokenType tokenType, int depth, bool hasConsumedAnyBytes, ref Utf8JsonReader reader)
+        {
+            switch (tokenType)
+            {
+                case JsonTokenType.StartArray:
+                    if (reader.TokenType != JsonTokenType.EndArray)
+                    {
+                        ThrowHelper.ThrowJsonException_SerializationConverterRead(this);
+                    }
+                    else if (depth != reader.CurrentDepth)
+                    {
+                        ThrowHelper.ThrowJsonException_SerializationConverterRead(this);
+                    }
```

Comment:
This logic was not changed by this PR. The conditions are separate to help with debugging since it may be useful to determine which path was taken. An alternative is to use separate exception messages for tokentype vs depth.

Reply:
> The conditions are separate to help with debugging since it may be useful to determine which path was taken

When debugging, you have all the state information so you could quite easily figure out what condition trigger it. Not sure, how separating it out helps. That said, keeping it unchanged is good since that's what master looks like.

Reply:
- It is possible to easily set a breakpoint.
- The exception will have a line number which could be used to troubleshoot